api/envoy/extensions/quic/server_preferred_address/v3/fixed_server_preferred_address_config.proto

```diff
@@ -16,20 +18,49 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#extension: envoy.quic.server_preferred_address.fixed]
 
 // Configuration for FixedServerPreferredAddressConfig.
+// [#next-free-field: 7]
 message FixedServerPreferredAddressConfig {
   // [#comment:TODO(danzh2010): discuss with API shepherds before removing WiP status.]
 
   option (xds.annotations.v3.message_status).work_in_progress = true;
 
+  // Addreses for server preferred address for a single address family (IPv4 or IPv6).
+  message Addresses {
+    // The server preferred address sent to clients.
+    //
+    // Note: Envoy currently must receive all packets for a QUIC connection on the same port, so
+    // unless `dnat_address` is configured, the port for this address must be zero, and the listener's
+    // port will be used instead.
+    config.core.v3.SocketAddress address = 1;
+
+    // If there is a DNAT between the client and Envoy, the address that Envoy will observe
+    // server preferred address packets being sent to. If this is not specified, it is assumed
+    // there is no DNAT and the server preferred address packets will be sent to the address advertised
+    // to clients for server preferred address.
+    //
+    // Note: Envoy currently must receive all packets for a QUIC connection on the same port, so the
+    // port for this address must be zero, and the listener's port will be used instead.
+
+    config.core.v3.SocketAddress dnat_address = 1;
+  }
+
   oneof ipv4_type {
     // String representation of IPv4 address, i.e. "127.0.0.2".
+    // The listener's port will be used.
     // If not specified, none will be configured.
     string ipv4_address = 1;
+
+    // The IPv4 address to advertise to clients for Server Preferred Address.
+    Addresses ipv4_config = 3;
   }
 
   oneof ipv6_type {
     // String representation of IPv6 address, i.e. "::1".
+    // The listener's port will be used.
     // If not specified, none will be configured.
     string ipv6_address = 2;
+
+    // The IPv6 address to advertise to clients for Server Preferred Address.
+    Addresses ipv6_config = 4;
```

Comment:
The structure of all this LGTM.  

I'm unsure if we should try to move the new `Address` type fields out of the `oneof` though, in keeping with the new API style guidance discouraging `oneof` in favor of defining precedence rules (e.g. this PR does that for an existing API https://github.com/envoyproxy/envoy/pull/33749/files#diff-e18722c3190aeb921472834ba2f4725bcfa5dfae5c024d9d7aca8cd2ecefeaee).

That would mean keeping the `oneof`, but defining `ipv4_config` and `ipv6_config` as separate fields outside of the `oneof` and using precedence rules that if ipv6_config (or ipv4_config) is defined, use that, otherwise use ipv6_address (or ipv4_address).  

Would that work?

Reply:
Pushed another version, which I think changes it as you're requesting. PTAL.

Reply:
Are any of the normal rules for changing protos relaxed due to `option (xds.annotations.v3.message_status).work_in_progress = true;`? Could we remove the oneof wrapper, since nothing else will ever go into it?

Reply:
LGTM, thanks Greg!

If it's marked work_in_progress, we could remove the `oneof` AFAIU.  In https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md, we have this clause:

```
Use a (xds.annotations.v3.file_status).work_in_progress, (xds.annotations.v3.message_status).work_in_progress, or (xds.annotations.v3.field_status).work_in_progress option annotation for files, messages, or fields, respectively, that are considered work in progress and are not subject to the threat model or the breaking change policy.
```

So we should be able to change the `oneof` since it's marked as work_in_progress. 

cc @envoyproxy/api-shepherds in case there is a differing opinion on this.

Reply:
The current structure looks good, other than a nits for renaming. Having `oneof` also makes sense to me, is there a reason why we want to remove it?

Reply:
`oneof` is now discouraged by the API style guide: https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md?plain=1#L63-L90

Reply:
@danzh2010 I didn't want to change the names to be overly specific in case we need to add new fields, and they don't work well with those very specific names. For `Addresses`, maybe is could be `AddressFamilyConfig`?

Reply:
AddressFamilyConfig sounds better.

Reply:
Changed name to `AddressFamilyConfig`